File: engine/scripts/gen_frag_tex.py
#!/usr/bin/env python

import sop, sys
from string import Template

the_chapter = sys.argv[1]
the_md = sys.argv[2]

chapter_meta = sop.get_chapter_meta(the_md)

authors = chapter_meta["author"]
if type(authors) is list:
    if len(authors) > 1:
        last_author = authors.pop()
        authors = ', '.join(authors) + " \\& " + last_author
    else:
        authors = authors[0]

tex = "\\chapter{$title}\n\n\\chapterauthor{$authors}\n\n\\input{$chapter-body}\n"

if "bibliography" in chapter_meta:
    tex += "\n\\bibliographystyle{plainnat}\n\\bibliography{$chapter}\n"

print Template(tex).safe_substitute(chapter = the_chapter,
                                    title = chapter_meta["title"],
                                    authors = authors)

File: engine/scripts/gen_frag_tex.R
#!/usr/bin/env Rscript

the_md <- commandArgs(trailingOnly = TRUE)[1]

read_frontmatter <- function(file) {
  file_content <- readLines(file)
  delimiters <- grep("^(---|\\.\\.\\.)\\s*$", file_content)
  frontmatter <- paste(file_content[(delimiters[1] + 1):(delimiters[2] - 1)], collapse = "\n")
  yaml::yaml.load(frontmatter)
}

chapter_meta <- read_frontmatter(the_md)
chapter_name <- tools::file_path_sans_ext(basename(the_md))

chapter_authors <- chapter_meta$author
if (length(chapter_authors) > 1L) {
  chapter_authors <- paste0(paste0(chapter_authors[-length(chapter_authors)], collapse = ", "),
                            " \\& ",
                            chapter_authors[length(chapter_authors)])
}

cat(sprintf("\\chapter{%s}\n\n\\chapterauthor{%s}\n\n\\input{%s-body}\n",
            chapter_meta$title, chapter_authors, chapter_name))

if ("bibliography" %in% names(chapter_meta)) {
  cat(sprintf("\n\\bibliographystyle{plainnat}\n\\bibliography{%s}\n",
              chapter_name))
}
